File: testing/runner.cpp
// testing/runner.cpp
#include <iostream>
#include <fstream>
#include <yaml-cpp/yaml.h>
#include <cstdlib>
#include <filesystem>

int main() {
    const std::string configPath = "testing/tests.yaml";
    if (!std::filesystem::exists(configPath)) {
        std::cerr << "❌ Config not found: " << configPath << std::endl;
        return 1;
    }

    YAML::Node config = YAML::LoadFile(configPath);
    if (!config["tests"]) {
        std::cerr << "❌ No 'tests' key in " << configPath << std::endl;
        return 1;
    }

    for (const auto& item : config["tests"]) {
        std::string project = item.first.as<std::string>();
        std::cout << "\n🚀 Running tests for: " << project << std::endl;

        for (const auto& fileNode : item.second) {
            std::string file = fileNode.as<std::string>();
            if (!std::filesystem::exists(file)) {
                std::cerr << "❌ File not found: " << file << std::endl;
                return 1;
            }

            std::string cmd = "node \"" + file + "\"";
            std::cout << "▶️  Executing: " << cmd << std::endl;
            int result = std::system(cmd.c_str());
            if (result != 0) {
                std::cerr << "❌ Test failed in " << file << std::endl;
                return result;
            }
        }
    }

    std::cout << "\n✅ All tests passed successfully.\n";
    return 0;
}

File: testing/runner.js
// testing/runner.js
import fs from "fs";
import { execSync } from "child_process";
import YAML from "yaml";
import chalk from "chalk";

// Config path
const configPath = "testing/tests.yaml";

// Color helpers
const log = {
  info: (...msg) => console.log(chalk.cyan(...msg)),
  warn: (...msg) => console.warn(chalk.yellow(...msg)),
  error: (...msg) => console.error(chalk.red(...msg)),
  success: (...msg) => console.log(chalk.green(...msg)),
};

// Check config file
if (!fs.existsSync(configPath)) {
  log.error(`❌ Config not found: ${configPath}`);
  process.exit(1);
}

// Load YAML
let config;
try {
  const file = fs.readFileSync(configPath, "utf8");
  config = YAML.parse(file);
} catch (err) {
  log.error(`❌ Failed to load YAML: ${err.message}`);
  process.exit(1);
}

// Validate tests key
if (!config.tests || typeof config.tests !== "object") {
  log.error(`❌ No valid 'tests' key in ${configPath}`);
  process.exit(1);
}

let allPassed = true;

for (const [projectName, files] of Object.entries(config.tests)) {
  console.log(`\n🚀 Running tests for: ${chalk.cyan(projectName)}`);

  if (!Array.isArray(files)) {
    log.error(`❌ Expected a list of files for project: ${projectName}`);
    allPassed = false;
    continue;
  }

  for (const filePath of files) {
    if (!fs.existsSync(filePath)) {
      log.error(`❌ File not found: ${filePath}`);
      allPassed = false;
      continue;
    }

    const cmd = `npx mocha "${filePath}"`;
    console.log(`▶️  Executing: ${chalk.yellow(cmd)}`);

    try {
      execSync(cmd, { stdio: "inherit" });
    } catch {
      log.error(`❌ Test failed in ${filePath}`);
      allPassed = false;
    }
  }
}

if (allPassed) {
  log.success("\n✅ All tests passed successfully.");
  process.exit(0);
} else {
  log.error("\n❌ Some tests failed.");
  process.exit(1);
}
